pkg/trace/event/extractor_legacy.go

```diff
@@ -23,8 +24,8 @@ func NewLegacyExtractor(rateByService map[string]float64) Extractor {
 // span's service. In this case the extracted event is returned along with the found extraction rate and a true value.
 // If this rate doesn't exist or the provided span is not a top level one, then no extraction is done and false is
 // returned as the third value, with the others being invalid.
-func (e *legacyExtractor) Extract(s *stats.WeightedSpan, priority sampler.SamplingPriority) (float64, bool) {
-	if !s.TopLevel {
+func (e *legacyExtractor) Extract(s *pb.Span, priority sampler.SamplingPriority) (float64, bool) {
+	if !traceutil.HasTopLevel(s) {
```

Comment:
This has the main difference of reading from `s.Metrics` which might be the source of a race condition.

Are we certain that no other goroutine doing the same (trace sampler? concentrator?) might write to it at the same time?

Reply:
That's a very good thought, but I can't see any issues:
* The event sampler runs in [the same goroutine](https://github.com/DataDog/datadog-agent/blob/0fda60bc8d9cca809740da287d7933c7b4a3f5e2/cmd/trace-agent/agent.go#L262-L283) as the sampler, sequentially.
* Looking at `(*stats.Concentrator).Add` and the entire `stats` package I can't find anyone writing to the span's `Metrics` field, besides [`SetSublayersOnSpan`](https://github.com/DataDog/datadog-agent/blob/0fda60bc8d9cca809740da287d7933c7b4a3f5e2/pkg/trace/stats/sublayers.go#L266) which [gets called](https://github.com/DataDog/datadog-agent/blob/0fda60bc8d9cca809740da287d7933c7b4a3f5e2/cmd/trace-agent/agent.go#L226) before any of the goroutines are started.

As far as I can see we should be good to go.

Reply:
There are also the `BenchmarkAgentTraceProcessing*` set of benchmarks which should most likely reveal a race condition like this when run with the `-race` flag.

Nevertheless, it feels like it's easy to introduce race conditions into the agent without realizing. I'll make sure to make this more difficult either via restructuring some code or adding even more tests (and continuing to avoid locking mechanisms).

Reply:
Good, my main concern was the sublayers and the trace sampler. But it looks like we already addressed that race condition in the past!

Reply:
Thanks for checking.